module/input/SensorFilter/src/SensorFilter.cpp

```diff
@@ -466,63 +487,76 @@ namespace module::input {
-                                             config.motionFilter.noise.measurement.flatFootOdometry,
-                                             MeasurementType::FLAT_FOOT_ODOMETRY());
-
-                        // do a foot based orientation update
-                        Eigen::Quaterniond Rwt(footlanding_Hwt.linear());
-                        motionFilter.measure(Rwt.coeffs(),
-                                             config.motionFilter.noise.measurement.flatFootOrientation,
-                                             MeasurementType::FLAT_FOOT_ORIENTATION());
+                        // If both feet are on the ground then we don't need to do another update
+                        if (!update_done) {
+                            // The difference between our current and previous Foot to torso CoM vectors is how much our
+                            // torso has moved in the last time step in torso space
+                            // We need to rotate this into world space to update our current Torso CoM position
+                            const Eigen::Quaterniond& Rwt     = MotionModel<double>::StateVec(motionFilter.get()).Rwt;
+                            const Eigen::Vector3d rMFt_update = current_rMFt - rMFt[side];
+                            const Eigen::Quaterniond q(0.0, rMFt_update.x(), rMFt_update.y(), rMFt_update.z());
+                            rTWw += (Rwt * q * Rwt.conjugate()).vec();
+
+                            // Make sure we don't do another update
+                            update_done = true;
+                        }
+                        previous_foot_down[side] = true;
+                        rMFt[side]               = current_rMFt;
                     }
-                    else if (!foot_down) {
+                    else {
                         previous_foot_down[side] = false;
                     }
-
-                    sensors->feet[side].Hwf = footlanding_Hwf[side].matrix();
                 }
 
-                // Calculate our time offset from the last read
-                double deltaT = std::max(
-                    (input.timestamp - (previousSensors ? previousSensors->timestamp : input.timestamp)).count()
-                        / double(NUClear::clock::period::den),
-                    0.0);
-
-                // Time update
-                motionFilter.time(deltaT);
-
+                /************************************************
+                 *       Construct Odometry Output (Htw)        *
+                 ************************************************/
                 // Gives us the quaternion representation
                 const auto o = MotionModel<double>::StateVec(motionFilter.get());
 
                 // Map from world to torso coordinates (Rtw)
                 Eigen::Affine3d Hwt;
                 Hwt.linear()      = o.Rwt.toRotationMatrix();
-                Hwt.translation() = o.rTWw;
+                Hwt.translation() = rTWw;
```

Comment:
For now i'd leave the calculations in here, I'd like to be able to see which one is more reliable and possibly fuse them. But we can use this for now
``` suggestion
Hwt.translation() = o.rTWw;
```